File: src/main/org/nlogo/swing/OptionDialog.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.swing;

import java.util.Arrays;
import java.util.List;

public strictfp class OptionDialog<T>
    extends UserDialog {
  javax.swing.JComboBox<T> options;

  public OptionDialog(java.awt.Frame owner, String title,
                      String message, T[] choices, scala.Function1<String, String> i18n) {
    super(owner, title, i18n);
    options = new javax.swing.JComboBox<T>(choices);
    addComponents(options, message);
  }

  public Object showOptionDialog() {
    java.awt.Rectangle r = getOwner().getBounds();
    setLocation
        (r.x + (r.width / 2) - (getWidth() / 2),
            r.y + (r.height / 2) - (getHeight() / 2));
    setVisible(true);
    if (selection == 0) {
      return Integer.valueOf(options.getSelectedIndex());
    } else {
      return null;
    }
  }

  public static int show(java.awt.Component owner, String title,
                         String message, Object[] options) {
    List<String> brokenLines = org.nlogo.awt.LineBreaker.breakLines
        (message,
            owner.getFontMetrics(owner.getFont()), DIALOG_WIDTH);
    StringBuilder brokenMessage = new StringBuilder();
    for (int i = 0; i < brokenLines.size(); i++) {
      brokenMessage.append(brokenLines.get(i));
      if (i + 1 < brokenLines.size()) {
        brokenMessage.append("\n");
      }
    }
    return javax.swing.JOptionPane.showOptionDialog
        (org.nlogo.awt.Hierarchy.getFrame(owner),
            brokenMessage.toString(), title,
            javax.swing.JOptionPane.DEFAULT_OPTION,
            javax.swing.JOptionPane.QUESTION_MESSAGE,
            null, options, options[0]);
  }

  public static int showIgnoringCloseBox(java.awt.Component owner, String title,
                                         String message, Object[] options, boolean asList) {
    int result = -1;
    while (result == -1) {
      if (asList) {
        result = showAsList(owner, title, message, options);
      } else {
        result = show(owner, title, message, options);
      }
    }
    return result;
  }

  public static int showAsList(java.awt.Component owner, String title,
                               String message, Object[] options) {
    return Arrays.asList(options).indexOf
        (javax.swing.JOptionPane.showInputDialog
            (org.nlogo.awt.Hierarchy.getFrame(owner),
                message, title,
                javax.swing.JOptionPane.QUESTION_MESSAGE,
                null, options, options[0]));
  }

}

File: netlogo-gui/src/main/swing/OptionDialog.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.swing

import java.awt.{ Component, Frame }
import javax.swing.{ JComboBox, JOptionPane }

import org.nlogo.awt.{ Hierarchy, LineBreaker }

@deprecated("use OptionPane instead for theme synchronization", "7.0")
class OptionDialog[T <: AnyRef](owner: Frame, title: String, message: String, choices: Array[T], i18n: String => String)
extends UserDialog(owner, title, i18n) {
  private val options = new JComboBox[T](choices)
  addComponents(options, message)

  def showOptionDialog(): AnyRef = {
    val r = getOwner.getBounds
    setLocation(r.x + (r.width  / 2) - (getWidth  / 2),
                r.y + (r.height / 2) - (getHeight / 2))
    setVisible(true)
    if (selection == 0) options.getSelectedIndex: Integer else null
  }
}

object OptionDialog {
  def showMessage(owner: Component, title: String, message: String, options: Array[? <: Object]): Int = {
    val brokenLines = LineBreaker.breakLines(message,
      owner.getFontMetrics(owner.getFont), UserDialog.DIALOG_WIDTH)
    showCustom(owner, title, brokenLines.mkString("\n"), options)
  }

  def showCustom(owner: Component, title: String, message: AnyRef, options: Array[? <: Object]): Int = {
    val parent: Component = Hierarchy.getWindow(owner)
    JOptionPane.showOptionDialog(parent, message, title,
      JOptionPane.DEFAULT_OPTION, JOptionPane.QUESTION_MESSAGE, null,
      options.asInstanceOf[Array[Object]], options(0))
  }

  def showIgnoringCloseBox(owner: Component, title: String, message: String, options: Array[AnyRef], asList: Boolean) = {
    val showDialog = {
      val showFunction = if (asList) showAsList else showMessage
      () => showFunction(owner, title, message, options)
    }
    var result = -1
    while (result == -1)
      result = showDialog()
    result
  }

  def showAsList(owner: Component, title: String, message: String, options: Array[AnyRef]) =
    options.indexOf(JOptionPane.showInputDialog(
      Hierarchy.getFrame(owner), message, title,
      JOptionPane.QUESTION_MESSAGE, null, options, options(0)))
}
